data/preprocess/util.py

```diff
@@ -0,0 +1,362 @@
+import numpy as np
+from os import path
+from dataclasses import dataclass, field
+from typing import Any, Callable, Dict, List, Optional, Union, Iterable
+import yaml
+
+### DEFAULT PREPROCESS UTILS ##########################################################
+
+
+@dataclass
+class RawSplitBase:
+    """Base RawSplit, tracks data and meta data of a raw split.
+
+    Attributes:
+        file (str): File with tab-separated raw triples (can be labeled)
+        collect_entities (bool): If true, entities contained in this split will be
+            collected during preprocessing.
+        collect_relations (bool): If true, relations contained in this split will be
+            collected during preprocessing.
+        raw_data (list): List of triples represented with raw ids; determined during
+            preprocessing
+        size (str): Number of triples; determined during preprocessing.
+        entities (dict): Entities contained in this split if collected. Keys refer to
+            raw id's and values to a dense index assigned during pre-processing.
+        relations (dict): See entities.
+
+    """
+
+    # fields defined by user
+    file: str
+    collect_entities: bool = False
+    collect_relations: bool = False
+
+    # fields defined during pre-processing
+    raw_data: List[str] = None
+    order_sop: bool = False
+    size: int = None
+
+    entities: Dict = field(default_factory=dict)
+    relations: Dict = field(default_factory=dict)
+
+    def write_splits(self, entities: Dict, relations: Dict, folder: str):
+        """Write the derived splits of this RawSplit and collect meta-data. """
+        raise NotImplemented
+
+    def update_config(self, config: Dict) -> Dict:
+        """Update a dataset config with meta data of the derived splits."""
+        raise NotImplemented
+
+
+@dataclass
+class RawSplit(RawSplitBase):
```

Comment:
"Raw" confused me first because "raw" implies that further processing has to be done to the split, i.e. its not the final split, isn't just a "Split"?